napari/_qt/qt_viewer.py

```diff
@@ -655,78 +655,71 @@ def on_resize(self, event):
         """
         self.viewer._canvas_size = tuple(self.canvas.size[::-1])
 
-    def on_mouse_wheel(self, event):
-        """Called whenever mouse wheel activated in canvas.
-
+    def _process_mouse_event(self, mouse_callbacks, event):
+        """Called whenever mouse pressed in canvas.
         Parameters
         ----------
-        event : qtpy.QtCore.QEvent
+        mouse_callbacks : function
+            Mouse callbacks function.
+        event : vispy.event.Event
+            The vispy event that triggered this method.
         """
         if event.pos is None:
             return
 
-        event = ReadOnlyWrapper(event)
+        # Update the cursor position
         self.viewer.cursor.position = self._map_canvas2world(list(event.pos))
-        mouse_wheel_callbacks(self.viewer, event)
+
+        # Add the cursor position to the event
+        event.position = self.viewer.cursor.position
+
+        # Put a read only wrapper on the event
+        event = ReadOnlyWrapper(event)
```

Comment:
this is also peripheral to this PR: but I'm not convinced we need to be wrapping all of these events with the `ReadOnlyWrapper`.  It kinda feels overly-cautious, and adds object creation overhead to a _very_ frequent thing.

Reply:
Yeah that's fair, though I also don't mind it as the event does get persisted through multiple callbacks. To be honest I think most events should really be read only by default, but maybe that's a discussion for another time. Right now thought it's actually not possible to just drop because of how persistence is done through the generator, so it would require more work. Happy to flag in a followup issue.

Reply:
> because of how persistence is done through the generator

can you elaborate?

I agree with @tlambert03's concerns though, this happens on each mouse move, it'd be at least interesting to measure how long creating the ROW takes.

Reply:
It takes around 3us which isn't too bad, given at 60Hz frame rate is ~16.6ms.

The event gets "saved" here https://github.com/napari/napari/blob/c2b254f9b7a54f1256f0bb853ec7875c9deea4bd/napari/utils/interactions.py#L145-L146, it's a little complex. We should probably simplify.